### ISSUE TITLE
RX 570 on Fedora 33 (Linux 5.11 w/o dkms): "hipErrorNoBinaryForGpu: Unable to find code object for all current devices!"

### ISSUE THREAD

**Issue by darkbasic** [ROCmSoftwarePlatform/tensorflow-upstream#1305]
**System information**
- OS Platform and Distribution: Fedora 33, Linux 5.11.8-200 (no DKMS)
- TensorFlow installed from: rpm repository
- TensorFlow version: v2.4.0-rc4-86-gd9cfe09e0d4 2.4.0
- Python version: Python 3.8.8 (venv)
- GPU model and memory: RX 570

**Describe the current behavior**

`python3.8 -m venv tensorflow`
`. tensorflow/bin/activate`
`pip install tensorflow-rocm --upgrade`
`git clone https://github.com/tensorflow/benchmarks.git`
`python3 benchmarks/scripts/tf_cnn_benchmarks/tf_cnn_benchmarks.py --num_gpus=1 --batch_size=32 --model=resnet50 --variable_update=parameter_server`
`deactivate`

```
WARNING:tensorflow:From /home/niko/venv/tensorflow/lib64/python3.8/site-packages/tensorflow/python/compat/v2_compat.py:96: disable_resource_variables (from tensorflow.python.ops.variable_scope) is deprecated and will be removed in a future version.
Instructions for updating:
non-resource variables are not supported in the long term
2021-03-25 19:03:07.999262: I tensorflow/core/platform/cpu_feature_guard.cc:142] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  SSE3 SSE4.1 SSE4.2 AVX AVX2 FMA
To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.
2021-03-25 19:03:07.999351: I tensorflow/compiler/jit/xla_gpu_device.cc:99] Not creating XLA devices, tf_xla_enable_xla_devices not set
2021-03-25 19:03:07.999508: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library libamdhip64.so
2021-03-25 19:03:08.020832: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1738] Found device 0 with properties: 
pciBusID: 0000:10:00.0 name: Ellesmere [Radeon RX 470/480/570/570X/580/580X/590]     ROCm AMD GPU ISA: gfx803
coreClock: 1.34GHz coreCount: 32 deviceMemorySize: 4.00GiB deviceMemoryBandwidth: 104.31GiB/s
2021-03-25 19:03:08.023187: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library librocblas.so
2021-03-25 19:03:08.024316: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library libMIOpen.so
2021-03-25 19:03:08.039284: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library librocfft.so
"hipErrorNoBinaryForGpu: Unable to find code object for all current devices!"
Fatal Python error: Aborted

Current thread 0x00007f2260717740 (most recent call first):
  File "/home/niko/venv/tensorflow/lib64/python3.8/site-packages/tensorflow/python/client/session.py", line 711 in __init__
  File "/home/niko/venv/tensorflow/lib64/python3.8/site-packages/tensorflow/python/client/session.py", line 1596 in __init__
  File "/home/niko/venv/benchmarks/scripts/tf_cnn_benchmarks/benchmark_cnn.py", line 3538 in setup
  File "benchmarks/scripts/tf_cnn_benchmarks/tf_cnn_benchmarks.py", line 61 in main
  File "/home/niko/venv/tensorflow/lib64/python3.8/site-packages/absl/app.py", line 251 in _run_main
  File "/home/niko/venv/tensorflow/lib64/python3.8/site-packages/absl/app.py", line 303 in run
  File "benchmarks/scripts/tf_cnn_benchmarks/tf_cnn_benchmarks.py", line 73 in <module>
Aborted (core dumped)
```

OpenCL works fine: https://openbenchmarking.org/result/2103244-AS-ROCM41DAR27

This is the first time I try to run anything which is not OpenCL with the ROCm stack, so I have no idea if other combinations of ROCm/kernel works.

**Describe the expected behavior**
Runs the benchmark.

**Other info / logs**

[tf_env.txt](https://github.com/ROCmSoftwarePlatform/tensorflow-upstream/files/6206848/tf_env.txt)

`rocminfo`:
```
ROCk module is loaded
=====================    
HSA System Attributes    
=====================    
Runtime Version:         1.1
System Timestamp Freq.:  1000.000000MHz
Sig. Max Wait Duration:  18446744073709551615 (0xFFFFFFFFFFFFFFFF) (timestamp count)
Machine Model:           LARGE                              
System Endianness:       LITTLE                             

==========               
HSA Agents               
==========               
*******                  
Agent 1                  
*******                  
  Name:                    AMD Ryzen 5 2400G with Radeon Vega Graphics
  Uuid:                    CPU-XX                             
  Marketing Name:          AMD Ryzen 5 2400G with Radeon Vega Graphics
  Vendor Name:             CPU                                
  Feature:                 None specified                     
  Profile:                 FULL_PROFILE                       
  Float Round Mode:        NEAR                               
  Max Queue Number:        0(0x0)                             
  Queue Min Size:          0(0x0)                             
  Queue Max Size:          0(0x0)                             
  Queue Type:              MULTI                              
  Node:                    0                                  
  Device Type:             CPU                                
  Cache Info:              
    L1:                      32768(0x8000) KB                   
  Chip ID:                 0(0x0)                             
  Cacheline Size:          64(0x40)                           
  Max Clock Freq. (MHz):   0                                  
  BDFID:                   0                                  
  Internal Node ID:        0                                  
  Compute Unit:            8                                  
  SIMDs per CU:            0                                  
  Shader Engines:          0                                  
  Shader Arrs. per Eng.:   0                                  
  WatchPts on Addr. Ranges:1                                  
  Features:                None
  Pool Info:               
    Pool 1                   
      Segment:                 GLOBAL; FLAGS: KERNARG, FINE GRAINED
      Size:                    65864980(0x3ed0514) KB             
      Allocatable:             TRUE                               
      Alloc Granule:           4KB                                
      Alloc Alignment:         4KB                                
      Accessible by all:       TRUE                               
    Pool 2                   
      Segment:                 GLOBAL; FLAGS: COARSE GRAINED      
      Size:                    65864980(0x3ed0514) KB             
      Allocatable:             TRUE                               
      Alloc Granule:           4KB                                
      Alloc Alignment:         4KB                                
      Accessible by all:       TRUE                               
  ISA Info:                
    N/A                      
*******                  
Agent 2                  
*******                  
  Name:                    gfx803                             
  Uuid:                    GPU-XX                             
  Marketing Name:          Ellesmere [Radeon RX 470/480/570/570X/580/580X/590]
  Vendor Name:             AMD                                
  Feature:                 KERNEL_DISPATCH                    
  Profile:                 BASE_PROFILE                       
  Float Round Mode:        NEAR                               
  Max Queue Number:        128(0x80)                          
  Queue Min Size:          4096(0x1000)                       
  Queue Max Size:          131072(0x20000)                    
  Queue Type:              MULTI                              
  Node:                    1                                  
  Device Type:             GPU                                
  Cache Info:              
    L1:                      16(0x10) KB                        
  Chip ID:                 26591(0x67df)                      
  Cacheline Size:          64(0x40)                           
  Max Clock Freq. (MHz):   1340                               
  BDFID:                   4096                               
  Internal Node ID:        1                                  
  Compute Unit:            32                                 
  SIMDs per CU:            4                                  
  Shader Engines:          4                                  
  Shader Arrs. per Eng.:   1                                  
  WatchPts on Addr. Ranges:4                                  
  Features:                KERNEL_DISPATCH 
  Fast F16 Operation:      FALSE                              
  Wavefront Size:          64(0x40)                           
  Workgroup Max Size:      1024(0x400)                        
  Workgroup Max Size per Dimension:
    x                        1024(0x400)                        
    y                        1024(0x400)                        
    z                        1024(0x400)                        
  Max Waves Per CU:        40(0x28)                           
  Max Work-item Per CU:    2560(0xa00)                        
  Grid Max Size:           4294967295(0xffffffff)             
  Grid Max Size per Dimension:
    x                        4294967295(0xffffffff)             
    y                        4294967295(0xffffffff)             
    z                        4294967295(0xffffffff)             
  Max fbarriers/Workgrp:   32                                 
  Pool Info:               
    Pool 1                   
      Segment:                 GLOBAL; FLAGS: COARSE GRAINED      
      Size:                    4194304(0x400000) KB               
      Allocatable:             TRUE                               
      Alloc Granule:           4KB                                
      Alloc Alignment:         4KB                                
      Accessible by all:       FALSE                              
    Pool 2                   
      Segment:                 GROUP                              
      Size:                    64(0x40) KB                        
      Allocatable:             FALSE                              
      Alloc Granule:           0KB                                
      Alloc Alignment:         0KB                                
      Accessible by all:       FALSE                              
  ISA Info:                
    ISA 1                    
      Name:                    amdgcn-amd-amdhsa--gfx803          
      Machine Models:          HSA_MACHINE_MODEL_LARGE            
      Profiles:                HSA_PROFILE_BASE                   
      Default Rounding Mode:   NEAR                               
      Default Rounding Mode:   NEAR                               
      Fast f16:                TRUE                               
      Workgroup Max Size:      1024(0x400)                        
      Workgroup Max Size per Dimension:
        x                        1024(0x400)                        
        y                        1024(0x400)                        
        z                        1024(0x400)                        
      Grid Max Size:           4294967295(0xffffffff)             
      Grid Max Size per Dimension:
        x                        4294967295(0xffffffff)             
        y                        4294967295(0xffffffff)             
        z                        4294967295(0xffffffff)             
      FBarrier Max Size:       32                                 
*** Done ***
```

**Comment by xuhuisheng** [comment]
The gfx803 series GPU - RX570, 580, wont get official support any more, since ROCm-4.0.
There is a unsolved bug on rocBLAS, since ROCm-3.7. Please refer here : https://github.com/RadeonOpenCompute/ROCm/issues/1265

You can downgrade to ROCm-3.5.1, ROCm-3.5.1 has some issues, too, but easy to sovle.

Or you can patch rocBLAS by yourself, here is ROCm-4.0 version https://github.com/xuhuisheng/rocm-build/tree/master/gfx803
I am working for ROCm-4.1 version.

**Comment by xuhuisheng** [comment]
Sorry, The bug comes from rocRAND, which is not support gfx803 on ROCm-4.1.
Here is the patch : https://github.com/ROCmSoftwarePlatform/rocRAND/commit/467824da5d312315dc97bf3590a1dd95ca0c6c55

And I belive you will meet the rocBLAS gfx803 bug next step.

**Comment by ROCmSupport** [comment]
Hi @darkbasic 
Thanks for reaching out.
We are not supporting gfx8 devices anymore with ROCm and so can not comment on this.
Please check for our hardware support section @ [https://github.com/RadeonOpenCompute/ROCm#hardware-and-software-support](url)
Thank you.

**Comment by darkbasic** [comment]
@ROCmSupport wow in a period of such GPU shortage you're going backwards instead, I have no words...

@xuhuisheng thanks but I use mostly OpenCL ATM, so I think I'll stick with the official packages. I feel sorry for those who I have advised to buy these cards though :(

**Comment by darkbasic** [comment]
@xuhuisheng is there anything preventing your fixes to be merged back into the upstream repo? What's the @ROCmSupport stance on this? I understand you won't officially support this, but this is an open source project and if the community wants to take the maintenance burden why stopping them to merge it back?

**Comment by xuhuisheng** [comment]
@darkbasic 
Actually, all of patches had been merged. They just cannot released with the current version of ROCm. Normally we will wait two versions to see them availiable.

ROCm had dropped official support for gfx803, so they won't test gfx803 on their develop environments. And following the release principle, When I start testing the release of new version, the ROCm team had frozon the features of version.

rocSPARSE case

1. Found rocSPARSE cannot support gfx803 on ROCm-3.9
2. Find the reason and make a PR https://github.com/ROCmSoftwarePlatform/rocSPARSE/pull/213
3. ROCm merged the PR
4. The patch had released in ROCm-4.1.

rocRAND case

1. Found rocBLAS cannot support gfx803 on ROCm-3.10
2. https://github.com/ROCmSoftwarePlatform/rocRAND/pull/159
3. ROCm merged the PR.
4. I guess this patch will release in ROCm-4.3.

In my opinion, the issues on missing gfx803 AMDGPU_TARGETS is easy. The invalid loss value of rocBLAS is hard to find the reason. It is random. And it  may be caused by imcompatible assembly codes and PCIe atomics op?

I am not familiar with GCN assembly or PCIe atomics op, and created an issue for tracing this. https://github.com/xuhuisheng/rocm-build/issues/4 . It is appreciate if anyone give any help. Thank you very much.

**Comment by darkbasic** [comment]
Thanks, I'm glad they're merging your patches. Unfortunately that won't solve the issue for anyone who use the binary packages. May I suggest to upload srpm packages to a copr in order to build a gfx803 compatible repo?

That would be extremely easy:
1) Create an account on https://copr.fedorainfracloud.org/ and name a new repo
2) Edit the srpm spec file to track a specific version of the source code
3) Build the packages for all the different versions of RHEL/Fedora on the cloud and automatically add them to your copr:
`copr-cli build --timeout 172800 rocm-gfx803 rocm-gfx803.src.rpm`

That would also allow to properly support Fedora and get rid of the following hacks:
`sudo sed -i 's/^#!.*/#!\/usr\/bin\/python/' /opt/rocm-4.1.0/bin/rocm_agent_enumerator`
`sudo sed -i 's/^#!.*/#!\/usr\/bin\/python/' /opt/rocm-4.1.0/bin/rocm_smi.py`

The biggest problem I can see is that @ROCmSupport doesn't publish the srpm for their packages, any reason why? Isn't ROCm supposed to be open source?
